File: lib/bin/test/evaluationtests/db/Orient.kt
package dbassistanttest.db


import com.beust.klaxon.Klaxon
import com.orientechnologies.orient.core.db.OrientDB
import com.orientechnologies.orient.core.db.OrientDBConfig
import com.orientechnologies.orient.core.db.OrientDBConfigBuilder
import com.orientechnologies.orient.core.db.document.ODatabaseDocument
import dbassistant.interfaces.IDBConnection

class Orient(
    val config: DBConfig
) : IDBConnection {

    private var orientDB: OrientDB? = null
    private var dbdoc: ODatabaseDocument? = null
    private val orientConfig: OrientDBConfig

    init{
        orientConfig = OrientDBConfigBuilder().build()        
    }

    override fun connect() {
        if(dbdoc == null || orientDB == null){
            orientDB = OrientDB(config.url, orientConfig)
            dbdoc = orientDB!!.open(config.db,config.user, config.password)
        }
    }

    override fun disconnect() {
        if(dbdoc != null){
            dbdoc!!.close()
            orientDB!!.close()
            dbdoc = null
            orientDB = null
        }
    }

    override fun query(query: String): List<Map<String,Any>>? {
        //SQL injection is absolutely possible, do not use in production
        //also parsing to json and back is very inefficient, but simple enough for the test
        if(dbdoc != null){
            val result = dbdoc!!.query(query)
            val set = mutableListOf<Map<String,Any>>()
            for (row in result){
                val map = Klaxon().parse<Map<String,Any>>(row.toJSON())
                if(map == null)throw Exception("Couldn't parse query result!")
                set.add(map)
            }
            return set
        }
        else{
            return null
        }
    }

}

File: example/Orient.py
import json
from pyorient import OrientDB


from pyorient import OrientSocket

class PySocket(OrientSocket):
    def __init__(self, host, port, protocol=38, timeout=300):
        super(PySocket, self).__init__(host, port)
        self.protocol = protocol
        self.host = host
        self.port = port
        self.timeout = timeout

    def connect(self):
        self._socket.settimeout(self.timeout)
        self._socket.connect( (self.host, self.port) )
        _value = self._socket.recv(2)

        if len(_value) != 2:
            self._socket.close()

        self.connected = True



class Orient:
    def __init__(self, config):
        self.config = config
        self.orientDB = None
        self.isOpen = False

    def connect(self):
        if self.isOpen is False or self.orientDB is None:
            socket = PySocket(self.config.url, 2424)
            socket.connect()
            self.orientDB = OrientDB(socket)
            self.orientDB.db_open(self.config.db, self.config.user, self.config.password)
            self.isOpen = True

    def disconnect(self):
        if self.isOpen is True and self.orientDB is not None:
            self.orientDB.db_close()
            self.orientDB = None
            self.isOpen = False

    def query(self, query):
        if self.isOpen is True and self.orientDB is not None:
            result = self.orientDB.query(query)
            data = []
            for row in result:
                data.append(row.oRecordData)
            return data
        else:
            return None
